Create incident handbook entry (#4051)
PostHog/posthog.com: contents/handbook/engineering/incidents.md

---

> title: Handling an incident

Reviewer: Note: we have already runbook to deal with incidents https://github.com/PostHog/product-internal/tree/main/infrastructure/runbooks we should likely consolidate those two.

Author: Done!

Reviewer: Did this happen?

Reviewer 2: I'm sorting this out once this PR has merged 👍

sidebar: Handbook
showTitle: true
---

Incidents are going to happen.

## When to raise an incident

**When in doubt, raise an incident.** We'd much rather have declared an incident which turned out not to be an incident. Many incidents take too long to get called, or are missed completely because someone didn't ring the alarm when they had a suspicion something was wrong.

To declare an incident, type `/incident` in the #incidents channel. This will create a new channel and send updates.

Anyone can declare an incident, including non-engineers. If in doubt, check with your nearest engineer.

Some things that should definitely be an incident

> - app.posthog.com being completely unavailable (not just for you)

Reviewer: ```suggestion
- `app.posthog.com` being completely unavailable (not just for you)
```

> - No insights can be created

Reviewer: Instead of explicitly calling what should be an incident, I would probably rephrase it as "every core functionality of the app works as designed"

Author: I'm calling out a couple of things specifically as examples as people may have a different definition of 'core functionality'

> - Feature flags are not being returned at all, or /decide is down

Reviewer: ```suggestion
- Feature flags are not being returned at all, or `/decide` is down
```

- Various alerts defined as critical, such as disk space full, OOM or >5 minute ingestion lag

Things that _shouldn’t_ be an incident
- Insights returning incorrect data
- Events being < 5 minutes behind

> - Unable to save insights, create feature flags

Reviewer: Can you expand why this should not be an incident?

Author: Saving insights doesn't stop you from using the insights, and not being able to create feature flags is annoying but doesn't break any existing functionality. Both should be treated as high priority bugs, but aren't incidents

## What happens during an incident

The person who raised the incident is the incident lead. It’s their responsibility to:

> - Make sure the right people join the zoom. This includes the current on call person (https://posthog.pagerduty.com/service-directory/P43Y0E8), people from Infra and the feature owner if relevant (https://posthog.com/handbook/engineering/feature-ownership)

Reviewer: Not sure how `incident.io` works but we should have a permanent Google Meet/Zoom room that people can bookmark and go there in case of issues.

Author: Incident.io does this automatically
<img width="562" alt="image" src="https://user-images.githubusercontent.com/1727427/186622075-3ea2f9ba-94a1-4840-87f2-202ed564ecb0.png">

> - Make sure the right people join the zoom. This includes the current on call person (https://posthog.pagerduty.com/service-directory/P43Y0E8), people from Infra and the feature owner if relevant (https://posthog.com/handbook/engineering/feature-ownership)

Reviewer: Is the expectation that infra folks will join _every_ incident? If yes, why?

Author: Genuine question, what kind of incidents do you think would not benefit/be resolved quicker if infra folks aren't there?

Reviewer: Any incident where whoever gets paged can quickly identify the issue and follow a runbook to fix it. 

Example:
- page for elevated overall error rate
- oncall person declares an incident 
- dashboard shows the error rate is coming from a specific endpoint
- oncall person identifies a PR related to that endpoint was recently deployed
- oncall person reverts the PR
- error rate goes back to nominal values 
- incident is declared mitigated
- once all follow up tasks are completed (eg notified the person/team responsible for the endpoint) the responder calls the "all clear"

Author: Make sense! I've made it "Optional if relevant" for both feature owner and platform folks.

- take notes in the incident channel. This should include time stamps, and is a brain dump of everything that we know, and everything that we are or have tried. This will give us much more opportunity to learn from the incident afterwards.

If the person who raised the incident is the best person to debug the issue, they should hand over the incident lead role to someone else on the call.

## When does an incident end?

> When we’re confident whatever was broken before is not broken anymore. This is sometimes hard to tell when the problem is intermittent. In that case, tell people to drop off the zoom, but to stay near their computer. Check the relevant metrics every 15 minutes or so, until you’re confident the problem is fully resolved. 

Reviewer: RC (root cause) is found and a fix is in place.

Author: I don't think every incident has a relevant fix? Sometimes with high loads there isn't much we can do but try to shed some load and wait

Reviewer: In this case I would consider load shedding as the fix

---
title: Handling an incident
sidebar: Handbook
showTitle: true
---

Incidents are going to happen.

## When to raise an incident

**When in doubt, raise an incident.** We'd much rather have declared an incident which turned out not to be an incident. Many incidents take too long to get called, or are missed completely because someone didn't ring the alarm when they had a suspicion something was wrong.

> To declare an incident, type `/incident` in the #incidents channel. This will create a new channel and send updates.

Reviewer 2: ```suggestion
To declare an incident, type `/incident` anywhere in Slack. This will create a new channel and send updates.
```

> 

Reviewer 2: I would also suggest adding guidance for choosing incident severity

eg

```
### Incident severity
Please refer to the following guidance when choosing the severity for your incident. If you are unsure, it's usually better to over-estimate than under-estimate! 

#### Minor
A minor-severity incident does not usually require paging people, and can be addressed within normal working hours. It is higher priority than any bugs however, and should come before sprint work.

Examples
- Broken non-critical functionality, with no workaround. Not on the critical path for customers
- Performance degradation. Not an outage, but our app is not performing as it should. For instance, growing (but not yet critical) ingestion lag.
- A memory leak in a database or feature. With time, this could cause a major/critical incident, but does not usually require _immediate_ attention

If not dealt with, minor incidents can often become major incidents. Minor incidents are usually OK to have open for a few days, whereas anything more severe we would be trying to resolve asap

#### Major
A major incident usually requires paging people, and should be dealt with _immediately_. They are usually opened when key or critical functionality is not working as expected.

Major incidents often become critical incidents if not resolved in a timely manner

Examples
- Customer signup is broken
- Significantly elevated error rate

#### Critical
An incident with very high impact on customers, and with the potential to existentially effect the company or reduce revenue

Examples
- Posthog Cloud is completely down
- A data breach, or loss of data
- Event ingestion totally failing - we are losing events
```

Anyone can declare an incident, including non-engineers. If in doubt, check with your nearest engineer.

Some things that should definitely be an incident
- `app.posthog.com` being completely unavailable (not just for you)
- No insights can be created
- Feature flags are not being returned at all, or `/decide` is down

> - Various alerts defined as critical, such as disk space full, OOM or >5 minute ingestion lag
> 

Reviewer 3: ```suggestion
- Various alerts defined as critical, such as disk space full, OOM or >5 minute ingestion lag
- Events not being ingested or ingestion being delayed significantly 

```
